### PR TITLE
Block inlining of IntroSort

### LINKED ISSUE #89106: [Perf] Windows/x64: 1 Regressions in System.Collections.Sort<IntStruct>
### Run Information

Name | Value
-- | --
Architecture | x64
OS | Windows 10.0.19042
Queue | OwlWindows
Baseline | [d59af2cf097acb100ad6a9cba652be34be6f4a2e](https://github.com/dotnet/runtime/commit/d59af2cf097acb100ad6a9cba652be34be6f4a2e)
Compare | [d9f6b98de38081b71fa9b5a463f302288546435d](https://github.com/dotnet/runtime/commit/d9f6b98de38081b71fa9b5a463f302288546435d)
Diff | [Diff](https://github.com/dotnet/runtime/compare/d59af2cf097acb100ad6a9cba652be34be6f4a2e...d9f6b98de38081b71fa9b5a463f302288546435d)
Configs | CompilationMode:tiered, RunKind:micro

### Regressions in System.Collections.Sort&lt;IntStruct&gt;

Benchmark | Baseline | Test | Test/Base | Test Quality | Edge Detector | Baseline IR | Compare IR | IR Ratio
-- | -- | -- | -- | -- | -- | -- | -- | --
[Array - Duration of single invocation](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_x64_Windows 10.0.19042/amd/System.Collections.Sort(IntStruct).Array(Size%3a%20512).html>)<br/>📝 - [Benchmark Source](<https://github.com/dotnet/performance/blob/main/src/benchmarks/micro/libraries/System.Collections/Sort.cs#L44-#L45>) | 5.03 μs | 6.10 μs | 1.21 | 0.39 | False | | |

graph
[Test Report](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/07_18_2023/refs/heads/main_x64_Windows%2010.0.19042/amd_Regression/System.Collections.Sort%28IntStruct%29.html>)

### Repro
General Docs link: https://github.com/dotnet/performance/blob/main/docs/benchmarking-workflow-dotnet-runtime.md

```cmd
git clone https://github.com/dotnet/performance.git
py .\performance\scripts\benchmarks_ci.py -f net8.0 --filter 'System.Collections.Sort&lt;IntStruct&gt;*'
```

<details>

### Payloads

[Baseline](<https://helixdi107v0xdeko0k025g8.blob.core.windows.net/helix-job-d161c67d-1114-4bf1-9b2e-2e0f49f26e2abf74077807a4fc48e/8e5e4c3b-6d85-4198-855f-c337dd14b27c.zip?sv=2021-08-06&se=2023-08-11T11%3A14%3A03Z&sr=c&sp=rl&sig=p9U7rTlctHnz7MFwyAldoHVBAiUnIAFLFvmkKCXfDR0%3D>)
[Compare](<https://helixdi107v0xdeko0k025g8.blob.core.windows.net/helix-job-cb44addd-cfc6-436f-90cb-f9c06e1bdd4214442b7e64349ad80/e0b308fa-220e-44e5-a915-877ea3f4b904.zip?sv=2021-08-06&se=2023-08-12T05%3A01%3A35Z&sr=c&sp=rl&sig=OLM2Mshx4B%2FnOylL3VrPdu6BPeF4I6KxHbXiWFYIEfc%3D>)
#### Description of detection logic
```
IsRegressionBase: Marked as regression because the compare was 5% greater than the baseline, and the value was not too small.
IsRegressionChecked: Marked as regression because the three check build points were 0.05 greater than the baseline.
IsImprovementBase: Marked as not an improvement because the compare was not 5% less than the baseline, or the value was too small.
IsRegressionBase: Marked as regression because the compare was 5% greater than the baseline, and the value was not too small.
IsRegressionChecked: Marked as regression because the three check build points were 0.05 greater than the baseline.
IsRegressionWindowed: Marked as regression because 6.098761428571428 > 5.2929065.
IsChangePoint: Marked as a change because one of 5/19/2023 4:22:12 AM, 7/12/2023 4:12:15 PM, 7/17/2023 10:23:26 PM falls between 7/8/2023 10:33:35 PM and 7/17/2023 10:23:26 PM.
IsRegressionStdDev: Marked as regression because -4.540498423685632 (T) = (0 -5812.810964285714) / Math.Sqrt((22347.782534576407 / (21)) + (172378.95924323285 / (8))) is less than -2.051830516474954 = MathNet.Numerics.Distributions.StudentT.InvCDF(0, 1, (21) + (8) - 2, .025) and -0.1330905133483964 = (5130.049979068551 - 5812.810964285714) / 5130.049979068551 is less than -0.05.
IsImprovementBase: Marked as not an improvement because the compare was not 5% less than the baseline, or the value was too small.
IsChangeEdgeDetector: Marked not as a regression because Edge Detector said so.

```
### Docs

[Profiling workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/profiling-workflow-dotnet-runtime.md)
[Benchmarking workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md)

</details>

### PR DESCRIPTION
**Pull request by AndyAyersMS**
With PGO and (via #88749) one level of recursive inlining enabled, the jit sees the recursive call made by `IntroSort` as an attractive inline candidate, but it isn't.

Fixes #89106.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@EgorBo @stephentoub PTAL
cc @dotnet/jit-contrib 

Benchmark is pretty noisy but this should resolve the regressions we've seen.

| Method |        Job |                                                                                 Toolchain | Size |     Mean |     Error |    StdDev |   Median |      Min |      Max | Ratio | RatioSD | Allocated | Alloc Ratio |
|------- |----------- |------------------------------------------------------------------------------------------ |----- |---------:|----------:|----------:|---------:|---------:|---------:|------:|--------:|----------:|------------:|
|  Array | Job-GINFRD |                                                       Pre-89106 |  512 | 4.969 us | 0.5558 us | 0.6177 us | 4.640 us | 4.253 us | 6.081 us |  1.00 |    0.00 |         - |          NA |
|  Array | Job-LXATAE |                                                       Post-89106 |  512 | 5.224 us | 0.6870 us | 0.7911 us | 4.966 us | 4.235 us | 6.730 us |  1.07 |    0.24 |         - |          NA |
|  Array | Job-LWDPQF | Noinline |  512 | 4.667 us | 0.6068 us | 0.6745 us | 4.440 us | 3.856 us | 6.299 us |  0.96 |    0.21 |         - |          NA |

File: src/libraries/System.Private.CoreLib/src/System/Array.cs [hunk 1]
```diff
@@ -2207,6 +2207,7 @@ private void IntrospectiveSort(int left, int length)
                 }
             }
 
+            [MethodImpl(MethodImplOptions.NoInlining)]
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Array.cs:2210]: Can we include a comment as to why this is here?  I imagine the calculus around whether this is valuable could change in the future.

**Review by stephentoub — APPROVED** [review]

**Review by EgorBo — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/Array.cs (hunk 1 above)
Comment (AndyAyersMS) [reply src/libraries/System.Private.CoreLib/src/System/Array.cs:2212]: Added. 

Note `IntroSort` is nominally doubly recursive but has already had the tail recursive call transformed into a loop manually. If we were ever tempted to undo that and rely on the JIT to do something similar, we might find the `NoInline` annotation would inhibit that transformation too.

@EgorBo we might consider special heuristics for recursion, generally if the method is large and doesn't have a frequent fast path that avoids recursion, then there's not much to be gained.

**Comment by AndyAyersMS** [comment]
Perf has returned to "normal" 
newplot (71)

Note there is a longer-term regression here from PGO, see #87194.
newplot (72)

**Pull request merged**